File: JavaScript/queue/queue.js
class Node {
    // Constructors
    constructor() {
        this._data = 0;
        this._next = null;
    }

    constructor(data) {
        this._data = data;
        this._next = null;
    }

    constructor(data, next) {
        this._data = data;
        this._next = next;
    }
}

class Queue {
    // Constructors
    constructor() {
        this._front = null;
        this._back = null;
        this._size = 0;
    }

    constructor(data) {
        this._front = Node(data);
        this._back = this._front;
        this._size = 1;
    }

    constructor(arr) {
        this._front = Node(arr[0]);
        this._back = this._front;
        this._size = 1;
        for (const element of arr) {
            this._back._next = Node(element);
            this._back = this._back._next;
            this._size++;
        }
    }

    // Public Methods
    enqueue(data) {
        if (this.isEmpty()) {
            this._front = Node(data);
            this._back = this._front;
            this._size++;
            return;
        }

        this._back._next = Node(data);
        this._back = this._back._next;
        this._size++;
    }

    dequeue() {
        if (this.isEmpty()) {
            console.log("Queue is empty");
            return;
        }

        if (this._size == 1) {
            this._front = null;
            this._back = null;
            this._size--;
            return;
        }

        this._front = this._front._next;
        this._size--;
    }

    peek() {
        if (this.isEmpty()) {
            console.log("Queue is empty");
            return -1;
        }

        return this._front._data;
    }

    isEmpty() {
        return this._size == 0;
    }

    getSize() {
        return this._size;
    }
}

File: Python/Queue/Queue.py
class Node:
    # Constructors
    def __init__(self):
        self._data = 0
        self._next = None

    def __init__(self, data):
        self._data = data
        self._next = None

    def __init__(self, data, next):
        self._data = data
        self._next = next


class Queue:
    # Constructors
    def __init__(self):
        self._front = None
        self._back = None
        self._size = 0

    def __init__(self, data):
        self._front = Node(data)
        self._back = self._front
        self._size = 1

    def __init__(self, list):
        self._front = Node(list[0])
        self._back = self._front
        self._size = 1

        for i in range(1, len(list)):
            self._back._next = Node(list[i])
            self._back = self._back._next
            self._size += 1

    # Methods
    def enqueue(self, data):
        if self._size == 0:
            self._front = Node(data)
            self._back = self._front
            self._size += 1
            return

        self._back._next = Node(data)
        self._back = self._back._next
        self._size += 1

    def dequeue(self):
        if self._size == 0:
            print("Queue is empty")
            return

        self._front = self._front._next
        self._size -= 1

    def peek(self):
        if self._size == 0:
            print("Queue is empty")
            return -1

        return self._front._data

    def get_size(self):
        return self._size

    def is_empty(self):
        return self._size == 0
